packages/vue-components/src/List.vue

```diff
@@ -0,0 +1,142 @@
+<template>
+  <component
+    :is="listType"
+    class="list-container"
+    :class="bulletStyle"
+  >
+    <div v-if="!parsedSlot || parsedSlot.length === 0">
+      <slot></slot>
+    </div>
+    <div
+      v-else
+    >
+      <ListItem
+        v-for="(item, index) in parsedSlot"
+        :key="index"
+        :icon="item.icon"
+      >
+        {{ item.content }}
+      </ListItem>
+    </div>
+  </component>
+</template>
+
+<script>
+import ListItem from './ListItem.vue';
+
+export default {
+  name: 'List',
+  components: {
+    ListItem,
+  },
+  provide() {
+    return {
+      list: this,
+    };
+  },
+  props: {
+    bulletStyle: {
+      type: String,
+      default: 'none',
+    },
+    bulletIcon: {
+      type: String,
+      default: '', // URL of the bullet icon image or FontAwesome class
+    },
+  },
+  computed: {
+    listType() {
+      const orderedListStyles = ['decimal', 'lower-alpha', 'upper-alpha', 'lower-roman', 'upper-roman'];
+      return orderedListStyles.includes(this.bulletStyle) ? 'ol' : 'ul';
+    },
+    parsedSlot() {
+      function adjustIndentation(code) {
+        const lines = code.split('\n');
+        let firstIndent = null;
+        const result = [];
+        let lastTopLevelLine = null;
+
+        lines.forEach((line) => {
+          if (line.trim() === '') return; // Skip empty lines
+          const lineIndent = line.match(/^\s*/)[0].length;
+          if (firstIndent === null) firstIndent = lineIndent;
+          const newLine = line.slice(firstIndent);
+          if (lineIndent > firstIndent) {
+            result[lastTopLevelLine] += `\n${newLine}`;
+          } else {
+            result.push(newLine);
+            lastTopLevelLine = result.length - 1;
+          }
+        });
+
+        return result;
+      }
+
+      const item = this.$slots.default[0];
+      if (!item.text) return undefined;
+
+      const text = item.text || '';
```

Comment:
> It is also important to maintain syntax consistency over the entire project.

Working with text alone may also reduce the feature set considerably. (what about HTML in list items?)

Reply:
This code only apply when the element is text. As demonstrated, you can use a combination of both text and HTML tags.

I will try to implement direct DOM manipulation after Markdown has processed the element.

Reply:
> This code only apply when the element is text. As demonstrated, you can use a combination of both text and HTML tags.

👍

> I will try to implement direct DOM manipulation after Markdown has processed the element.

Let's try to get an OK on the syntax first from @damithc, the implementation may change completely.

> direct DOM manipulation

https://github.com/MarkBind/markbind/pull/2309#discussion_r1237971902

Reply:
> Let's try to get an OK on the syntax first from @damithc, the implementation may change completely.
> 
> >

To confirm the following syntax:

```
<icon-list>

* {.icon=":fas-tick:"} first item
* second item
  * {.icon=":fas-x:" .class="text-danger"} third item
  * fourth item
</icon-list>
```

Here's the flow of the implementation:

1. The Markdown parser processes the 'icon-list'.
2. Based on the generated tags, we extract the 'icon' and 'class' attributes, then perform necessary CSS styling and DOM manipulation without the use of Vue components.

Reply:
My take is here https://github.com/MarkBind/markbind/pull/2309#discussion_r1237993636.

@damithc 

---

Implementation wise, the key part is performing the markdown on the backend. Some HTML transforms will have to be done there as well, to change the intermediate output directly into the final output or a vue component (whichever you find easier).